src/probability/kernel/composition.lean

```diff
@@ -0,0 +1,721 @@
+
+lemma comp_prod_eq_sum_comp_prod (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] :
+  κ ⊗ₖ η = kernel.sum (λ n, kernel.sum (λ m, seq κ n ⊗ₖ seq η m)) :=
+by { ext a s hs : 2, simp_rw [kernel.sum_apply' _ a hs], rw comp_prod_eq_tsum_comp_prod κ η a hs, }
+
+lemma comp_prod_eq_sum_comp_prod_left (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] :
+  κ ⊗ₖ η = kernel.sum (λ n, seq κ n ⊗ₖ η) :=
+begin
+  rw comp_prod_eq_sum_comp_prod,
+  congr' with n a s hs,
+  simp_rw [kernel.sum_apply' _ _ hs, comp_prod_apply_eq_comp_prod_fun _ _ _ hs,
+    comp_prod_fun_tsum_right _ η a hs],
+end
+
+lemma comp_prod_eq_sum_comp_prod_right (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] :
+  κ ⊗ₖ η = kernel.sum (λ n, κ ⊗ₖ seq η n) :=
+begin
+  rw comp_prod_eq_sum_comp_prod,
+  simp_rw comp_prod_eq_sum_comp_prod_left κ _,
+  rw kernel.sum_comm,
+end
+
+instance is_markov_kernel.comp_prod (κ : kernel α β) [is_markov_kernel κ]
+  (η : kernel (α × β) γ) [is_markov_kernel η] :
+  is_markov_kernel (κ ⊗ₖ η) :=
+⟨λ a, ⟨
+begin
+  rw comp_prod_apply κ η a measurable_set.univ,
+  simp only [set.mem_univ, set.set_of_true, measure_univ, lintegral_one],
+end⟩⟩
+
+lemma comp_prod_apply_univ_le (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_finite_kernel η] (a : α) :
+  (κ ⊗ₖ η) a set.univ ≤ (κ a set.univ) * (is_finite_kernel.bound η) :=
+begin
+  rw comp_prod_apply κ η a measurable_set.univ,
+  simp only [set.mem_univ, set.set_of_true],
+  let Cη := is_finite_kernel.bound η,
+  calc ∫⁻ b, η (a, b) set.univ ∂(κ a)
+      ≤ ∫⁻ b, Cη ∂(κ a) : lintegral_mono (λ b, measure_le_bound η (a, b) set.univ)
+  ... = Cη * κ a set.univ : lintegral_const Cη
+  ... = κ a set.univ * Cη : mul_comm _ _,
+end
+
+instance is_finite_kernel.comp_prod (κ : kernel α β) [is_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_finite_kernel η] :
+  is_finite_kernel (κ ⊗ₖ η) :=
+⟨⟨is_finite_kernel.bound κ * is_finite_kernel.bound η,
+  ennreal.mul_lt_top (is_finite_kernel.bound_ne_top κ) (is_finite_kernel.bound_ne_top η),
+  λ a, calc (κ ⊗ₖ η) a set.univ
+    ≤ (κ a set.univ) * is_finite_kernel.bound η : comp_prod_apply_univ_le κ η a
+... ≤ is_finite_kernel.bound κ * is_finite_kernel.bound η :
+        mul_le_mul (measure_le_bound κ a set.univ) le_rfl (zero_le _) (zero_le _), ⟩⟩
+
+instance is_s_finite_kernel.comp_prod (κ : kernel α β) [is_s_finite_kernel κ]
+  (η : kernel (α × β) γ) [is_s_finite_kernel η] :
+  is_s_finite_kernel (κ ⊗ₖ η) :=
+begin
+  rw comp_prod_eq_sum_comp_prod,
+  exact kernel.is_s_finite_kernel_sum (λ n, kernel.is_s_finite_kernel_sum infer_instance),
+end
+
+end composition_product
+
+section map_comap
+/-! ### map, comap -/
+
+variables {γ : Type*} {mγ : measurable_space γ} {f : β → γ} {g : γ → α}
+
+include mγ
+
+/-- The pushforward of a kernel along a measurable function. -/
```

Comment:
```suggestion
/-- The pushforward of a kernel along a measurable function. 
We include measurability in the assumptions instead of using junk values
to make sure that typeclass inference can infer that the `map` of a Markov kernel
is again a Markov kernel. -/
```
Did I understand the design correctly?

Reply:
To be honest, the "design" is accidental: I started by making a definition with all hypotheses needed to make it work and did not add junk values. But then that formulation proved really useful for typeclass inference, as you noted.
Right now I am writing code about composing chains of kernels (my goal is the Ionescu Tulcea theorem) and in order to deal with the type theory hell that comes from taking the product of finitely many types I use plenty of maps and comaps along measurable equivalences. I am very happy that inference can see through these.

I am still thinking about related design decisions: since every composition-like operation is defined only for s-finite kernels, should I define a type of s-finite kernels and work only with these?
For now my approach is to keep it unbundled and wait for the possible difficulties to arise.

Reply:
Actually, now that I though about it a bit more, it looks like changing `map` to take the junk value 0 if the function is not measurable could be a better idea. It would not change the fact that the map of a Markov kernel is a Markov kernel: that would be true regardless of the measurability of f. I'll experiment with that change and make a follow up PR if it's indeed better.